src/monitoring_tests/combinatorial_auction_surplus_test.py

```diff
@@ -0,0 +1,291 @@
+      the competition.
+    - Compute a baseline for surplus on all token pairs from those solutions.
+    - Filter solutions which do not provide at least as much surplus as the baseline on all token
+      pairs that the solution contains.
+    - Choose one batch winner and multiple single order winners.
+
+    The test logs an output whenever
+    - the winner of the solver competition would have been filtered out in a combinatorial auction
+      or
+    - the total surplus in a combinatorial auction would have been significantly larger than it was
+      with our current mechanism.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.web3_api = Web3API()
+        self.orderbook_api = OrderbookAPI()
+
+    def run_combinatorial_auction(self, competition_data: dict[str, Any]) -> bool:
+        # pylint: disable=too-many-locals
+        """Run combinatorial auction on competition data.
+
+        The combinatorial auction consists of 4 steps:
+        1. Aggregate surplus on the different directed sell token-buy token pairs for all solutions
+           in the competition.
+        2. Compute a baseline for surplus on all token pairs from those solutions.
+        3. Filter solutions which to not provide at least as much surplus as the baseline on all
+           token pairs.
+        4. Choose one batch winner and multiple single order winners.
+        """
+
+        solutions = competition_data["solutions"]
+        winning_solution = competition_data["solutions"][-1]
+
+        aggregate_solutions = [
+            self.get_token_pairs_surplus(
+                solution, competition_data["auction"]["prices"]
+            )
+            for solution in solutions
+        ]
+        winning_aggregate_solution = aggregate_solutions[-1]
+
+        baseline_surplus = self.compute_baseline_surplus(aggregate_solutions)
+        filter_mask = self.filter_solutions(aggregate_solutions, baseline_surplus)
+        winning_solutions = self.determine_winning_solutions(
+            aggregate_solutions, baseline_surplus, filter_mask
+        )
+        winning_solvers = self.determine_winning_solvers(
+            solutions, aggregate_solutions, winning_solutions
+        )
+
+        solutions_filtering_winner = [
+            solution["solver"]
+            for i, solution in enumerate(solutions)
+            if i in filter_mask[-1]
+        ]
+
+        total_combinatorial_surplus = sum(
+            sum(surplus for _, surplus in token_pair_surplus.items())
+            for _, token_pair_surplus in winning_solvers.items()
+        )
+        total_surplus = sum(
+            surplus for _, surplus in winning_aggregate_solution.items()
+        )
+
+        a_abs_eth = total_combinatorial_surplus - total_surplus
+
+        # convert surplus to float for logging
+        winning_aggregate_surplus_float = {
+            token_pair: float(surplus)
```

Comment:
Can we maybe reduce the number of decimals to 4? I guess we are not doing it other tests but it would make things more readable.

Something like `round(float_variable,4)` should do it.

Reply:
I added a more general function to process those dicts before logging.